### ISSUE TITLE
Spring cleaning in the materials/ folder

### ISSUE THREAD

**Issue by akien-mga** [OpenDungeons/OpenDungeons#579]
It is currently relatively hard to know which texture is linked to which material, and which material is used in game or not.

According to @oyvindln's input on #492, most material IDs (defined in the materials/script/*.material files by the `material` tag) are referenced directly in the binary meshes. The we reference the meshes in the config/ files.

So basically to know the status for each material/model, we can:
- Check that every texture is used in a material script
- Extract the material IDs from the scripts
- Check the meshes against those material IDs to know if they're using them
- Check the meshes against the config files (and the code if there are some hardcoded meshes) to know which are used

We don't necessarily have to do exactly this for each step, since the data folders are not so huge so for most models we know if we see them in game already.
I'm just trying to single out the textures/models that we might not be using actually.

**Comment by akien-mga** [comment]
For the first two points, I wrote a shell script that checks all textures files against the material scripts, and extracts the material IDs.

```
#!/bin/sh

search_source=false

pushd materials/textures > /dev/null
ls * > ../../../textures-list.txt
popd > /dev/null

for texture in $(cat ../textures-list.txt)
do
    echo "== Texture: $texture =="
    echo ". Files matching the filename:"
    ack -il $texture
    for script in $(ack -il $texture | grep materials/scripts)
    do
        line=$(cat $script | grep "^ *material " | sed -e 's/^ *material //')
        echo -e ". Matched material definition:\t$line"
        for word in $line
        do
            if [ -n "$(echo $word | grep RTSS)" -o -n "$(echo $word | grep :)" ]
            then
                continue
            fi
            echo -e ". Extracted material name:\t$word"
            if $search_source
            then
                echo ". References to $word in the source code:"
                pushd source > /dev/null
                ack -il $word
                popd > /dev/null
                echo ""
            fi
        done
    done
    echo ""
done
```

It takes a little white to process everything, but the output can be useful already to see that we have quite a few apparently unused textures.

**Comment by akien-mga** [comment]
Here is the output when running this script in the current development branch: https://gist.github.com/akien-mga/7c1d4825ee321d7ce4a3#file-texture-analyse-txt

**Comment by Bertram25** [comment]
Excellent :D This will definitely help!
I'd also add that models that aren't using materials should whenever possible. So tracking which model isn't using a material is also something to do and fix, right?

**Comment by Bertram25** [comment]
I vote for adding that script (with documentation header and copyright) into a tools/ folder for instance.
Other scripts like update copyright year and such could belong there.

**Comment by akien-mga** [comment]
From this I can single out a list of textures that seem to be used by no material script. Some of them are normal or specular maps that have not yet been linked to the corresponding model, so we shouldn't just blindly wipe them out, they will be relevant for #511 ;-)
- Axe.png
- big_diffus.png
- big_normal.png
- big_specular.png
- Cavehornet_nor.png
- Cavehornet_spec.png
- Claimedwall3.png
- Con_Worker.png
- CreatureOverlay.png
- Dirt.png
- DungeonTemple_glow.png
- Env-blaugruenF.png
- Flat_n.png
- Grey.png
- GrimoireNormal.png
- Halo2_0.png
- Halo2_1.png
- Halo2_2.png
- Halo2_3.png
- Keeperhand_colD.png
- Keeperhand_nor.png
- Keeperhand_spec.png
- Lava_anim_1.png
- Lava_anim_2.png
- Lava_anim_3.png
- Lava_anim_4.png
- Lizardman_nor3.png
- Lizardman_spec.png
- Obsidian.png
- Orc_nor.png
- Orc_spec.png
- Tentakel_diff_yellow7.png
- Tentakel_nor.png
- Treasury_specular.png
- Wizard_nor.png
- Wyvern_red_normalmap.png

**Comment by oyvindln** [comment]
Does it also check for unused materials?

**Comment by akien-mga** [comment]
> Does it also check for unused materials?

Not yet. As far as I know I would have to check that in the .mesh files, which are binary files. Should I try opening them in Blender, or is there another way to retrieve some metadata from them?

**Comment by oyvindln** [comment]
Grep might be able to find the names in the binary files. I don't think they are compressed. Alternatively they can be converted to xml files with ogrexmlconverter, which should at least have the material name somewhere.

**Comment by akien-mga** [comment]
> Grep might be able to find the names in the binary files. I don't think they are compressed.

It kind of works but it's tedious and would be hard to automatise.

> Alternatively they can be converted to xml files with ogrexmlconverter, which should at least have the material name somewhere.

Awesome! I tried OgreXMLConverter on Dragon.mesh and it gives exactly what I wanted to know :)

```
<submesh material="Dragon" usesharedvertices="false" use32bitindexes="false" operationtype="triangle_list">
```

I'll work on improving my script to first generate XML versions of all .mesh (+ .skeleton) and then run "ack" in those files to identify used materials.

**Comment by akien-mga** [comment]
> Awesome! I tried OgreXMLConverter on Dragon.mesh and it gives exactly what I wanted to know :)

Pro-tip for @hwoarangmy and @Bertram25: You can use OgreXMLConverter to convert .mesh and .skeleton files to XML, and this way find for example the name of existing animations without having to bother @Danimal696 :-)

**Comment by hwoarangmy** [comment]
I have used this tool to convert the xml file generated by blender into the mesh file while dealing with the keeper hand. FYI, there is a console command to list all animations and bones of every mesh ingame: listanims

BTW, I don't know if you are aware but I've downloaded a tool called Ogre meshy that allows to open and display a mesh and also to play its animations (nice when animations have exotic names) ^^

**Comment by akien-mga** [comment]
Here is the output from ./tools/resources-analysis.sh --scan-meshes: https://gist.github.com/akien-mga/d2817ddeeb7c47cb3dcb#file-mesh-analysis-txt

**Comment by akien-mga** [comment]
Analysis of the data will come later ;-)

**Comment by Bertram25** [comment]
It's cool! With this, we will easily track down unused textures/materials.
Is it possible to easily automate an ack on src/ and config/ for used model, for instance?

**Comment by akien-mga** [comment]
>  Is it possible to easily automate an ack on src/ and config/ for used model, for instance?

Sure, I did that at first but it could quickly because pretty messy for materials with generic names used in code functions, e.g. Treasury. I'll have to think a bit about how to make a pretty output.
I'm pretty happy about my trick to extract the material from the meshes though :-D

**Comment by akien-mga** [comment]
Here is the output from the feature added in https://github.com/akien-mga/OpenDungeons/commit/a8b42cc484e6bbacdbdb7aebe4888b3648cadcf8: https://framalab.org/zerobin/?c9e23f8ba7de7c76#/XpgI1JU598bbEXvHo5VtgZR+pB2RAmDY4WNL90ppHk=

My script is not the most efficient or readable implementation, but I'm mostly writing it to get experience with shell scripting :)

**Comment by hwoarangmy** [comment]
Can't you cross that with the files in material to check which are not used ?

**Comment by akien-mga** [comment]
> Can't you cross that with the files in material to check which are not used ?

Yeah good idea, I should probably do that in my "scan_meshes" function as I did for unused textures in the "scan_textures" part.

BTW on the output linked above we can see that we have some meshes referencing materials that are not in the repo, like "axt", "Fire/TWOSIDE" and "Torch/TWOSIDE" (search for "()" to find them).

**Comment by Bertram25** [comment]
@akien-mga Same here. The script seems to make this trivial to do.

**Comment by akien-mga** [comment]
I'm on it :)

**Comment by akien-mga** [comment]
I've had a look at the unused models to start. To try them in game I used them as BedMeshName for the BigKnight in the 2v2 map ;-) Here are my findings:
- Axe1.mesh: An axe, but with a bad texture it seems
  ![axe1](https://cloud.githubusercontent.com/assets/4701338/6866741/a4923ee0-d47b-11e4-9d34-4c7f1f0b2478.jpg)
- BeetleGolem.mesh: Obsoleted by Scarab.mesh I guess?
  ![beetlegolem](https://cloud.githubusercontent.com/assets/4701338/6866742/a49dc62a-d47b-11e4-924f-5de8d58b77e7.jpg)
- DojoRockObject.mesh: Same as Axe1.mesh?!
  ![dojorockobject](https://cloud.githubusercontent.com/assets/4701338/6866743/a694a1b0-d47b-11e4-8cb6-96ffd48929b3.jpg)
- Foundry.mesh: Recently added, will be used in 0.6.0
  ![foundry](https://cloud.githubusercontent.com/assets/4701338/6866785/09b7cdb2-d47c-11e4-823e-d8ca8ab93bc2.jpg)
- Hammock.mesh: Bed, not used by a creature yet, the material script needs love
  ![hammock](https://cloud.githubusercontent.com/assets/4701338/6866810/2cc607ec-d47c-11e4-9605-af642d8412f1.jpg)
- RuinEntrance.mesh: Not used
  ![ruinentrance](https://cloud.githubusercontent.com/assets/4701338/6866747/ab99a69c-d47b-11e4-8621-9e2729e42974.jpg)
- Torch.mesh: Not used for now, fire looks bad
  ![torch](https://cloud.githubusercontent.com/assets/4701338/6866750/ae856d1e-d47b-11e4-9831-1d2ac3bc5c37.jpg)
- Troll_Rock.mesh: Meant to be thrown by the golem?
  ![troll_rock](https://cloud.githubusercontent.com/assets/4701338/6866751/b0bfd966-d47b-11e4-849d-f884fe387319.jpg)
- WoodBridge.mesh: Recently added, will be used in 0.6.0
  ![woodbridge](https://cloud.githubusercontent.com/assets/4701338/6866776/e608eeb4-d47b-11e4-9ab5-ddeffe1af024.jpg)

**Comment by akien-mga** [comment]
IMO we should keep the ones that we intend to use soon, i.e.:
- Foundry.mesh
- Hammock.mesh (it needs to be attributed to a creature and its material script tweaked)
- Troll_Rock.mesh (if @Danimal696 confirms that we want to use it)
- WoodBridge.mesh

And drop:
- Axe1.mesh
- BeetleGolem.mesh
- DojoRockObject.mesh
- RuinEntrance.mesh
- Torch.mesh

**Comment by hwoarangmy** [comment]
> Troll_Rock.mesh (if @Danimal696 confirms that we want to use it)

When doing missile objects (bullets fired by the cannon), I've adapted everything so that creature can also fire objects. We can use it for the troll if we want. However, in the current battle logic, ranged creatures will always try to be at their maximum fire distance before firing. For such a slow creature as the troll, it would almost mean to never fire.
I guess that if we want the troll to be able to fire rocks, we should introduce a concept of ranged/melee/both creature that would try to fire like the others ranged creature when not attacked by a melee unit and that would fight melee if attacked by melee.

**Comment by Bertram25** [comment]
> I guess that if we want the troll to be able to fire rocks, we should introduce a concept of ranged/melee/both creature that would try to fire like the others ranged creature when not attacked by a melee unit and that would fight melee if attacked by melee.

Yep, IMHO, a good fix related to #376

**Comment by Danimal696** [comment]
Fixing #376 will be a need sooner or later when more ranged creatures are added, but i would let the rock issue for later when we try to implement creature skills.

Akien, you got your classification rigth, you can do as you want with the dropped ones

**Comment by akien-mga** [comment]
Ok, I'll move the "dropped" ones to the models_unused folder in the media repo.

**Comment by akien-mga** [comment]
There is another unused model that my script did not catch because it has a too generic name used in the code source:
- Fire.mesh: uses the "Fire/TWOSIDE" material which is not provided by any script

**Comment by Danimal696** [comment]
thats the old torch, its unused anymore

**Comment by akien-mga** [comment]
> thats the old torch, its unused anymore

That's what I guessed, so I dropped it too locally. I'll do a PR in a while.

**Comment by akien-mga** [comment]
I added a test in my script to list materials that are not used in mesh files:

```
[material name] [script filename]   [matched in source]
CreatureOverlay0    CreatureOverlay.material    0
CreatureOverlay1    CreatureOverlay.material    0
CreatureOverlay2    CreatureOverlay.material    0
CreatureOverlay3    CreatureOverlay.material    0
CreatureOverlay4    CreatureOverlay.material    0
GoldGround  GoldGround.material 8
golem_mat   LavaGolem_mat.material  0
Refl2   Sabre.material  0
pskmat1 WoodWatchTower_mat.material 0
```

Some are false positives because CreatureOverlay{1,2,3,4} are used in the source code directly as "CreatureOverlay + i", and my script searches for a 1:1 match so it doesn't find them.

GoldGround is used to configure the tileset, it doesn't need any mesh as far as I know since it's a ground tile, but I might have got it wrong when I created this material.

The other three can be removed I think.

**Comment by hwoarangmy** [comment]
> GoldGround is used to configure the tileset, it doesn't need any mesh as far as I know since it's a ground tile

It is used by ground mesh. The exact filename being defined in the tileset: Dirt_0.mesh

**Comment by akien-mga** [comment]
> >  GoldGround is used to configure the tileset, it doesn't need any mesh as far as I know since it's a ground tile
> 
> It is used by ground mesh. The exact filename being defined in the tileset: Dirt_0.mesh

You are right, I'm even the one who did that :-D

This shows that my scripts need to be handled with care, the results are useful but _absolutely_ not 100% accurate. Let's say they give a first list of potentially unused stuff that needs to be further investigated by checking again the mesh, the config files, the source, etc.

**Comment by akien-mga** [comment]
Ok I've done a first pass at removing unused models and unused materials.

Now I'm looking at the unused textures (according to my imperfect scripts, but that's still useful input). I've made a list with comments and I'd like to have @Danimal696's feedback on it so that I know what to remove and what to keep. Here it is:

> Axe.png: To drop, should have been removed with the other axe components
> big_diffus.png: ?
> big_normal.png: ?
> big_specular.png: ?
> Cavehornet_nor.png: For CaveHornet*.material
> Cavehornet_spec.png: For CaveHornet*.material
> Claimedwall3.png: To drop
> Con_Worker.png: Not used in Con_Worker.material?
> CreatureOverlay.png: Template for the other overlays? @hwoarangmy
> Dirt.png: To drop
> DungeonTemple_glow.png: For DungeonTemple.material
> Env-blaugruenF.png: To drop
> Flat_n.png: To drop, just a dot
> Grey.png: To drop, just a dot
> Halo2_0.png: ?
> Halo2_1.png: ?
> Halo2_2.png: ?
> Halo2_3.png: ?
> Keeperhand_colD.png: For Keeperhand.material
> Keeperhand_nor.PNG: For Keeperhand.material
> Keeperhand_spec.png: For Keeperhand.material
> Lava_anim_1.png: ?
> Lava_anim_2.png: ?
> Lava_anim_3.png: ?
> Lava_anim_4.png: ?
> Lizardman_nor3.png: For Lizardman.material
> Lizardman_spec.png: For Lizardman.material
> Obsidian.png: ?
> Orc_nor.png: For Orc.material
> Orc_spec.png: For Orc.material
> Tentakel_diff_yellow7.png: For Tentacle.material and/or Tentacle2.material
> Tentakel_nor.png: For Tentacle.material and/or Tentacle2.material
> Treasury_specular.png: For Treasury.material
> Wizard_nor.png: For Wizard.material
> Wyvern_red_normalmap.png: For Wyvern.material

So basically the ones with a _To drop_ comment should probably be dropped as they seem to be remains from a past long gone. The ones with _For something.material_ are probably meant to be used as normal or specular maps for existing models, if you see any in the list that you don't plan on using, please tell me. The ones with a _?_, I am not sure of their status. And the construct worker one seems unused, but shouldn't we remove the construct worker altogether?

**Comment by hwoarangmy** [comment]
Concerning CreatureOverlay.png, it is the original I used for the overlays. It can be dropped.

**Comment by Danimal696** [comment]
> big_diffus.png: ?
> big_normal.png: ?
> big_specular.png: ? 

Drop, those are for a barrel i have never seen?
Con_Worker.png -> Contruct worker could be removed, it doesnt fit at all now with OD
Con_gear -> same as above, but construct worker must be erased from some test levels then

> Halo2_0.png: ?
> Halo2_1.png: ?
> Halo2_2.png: ?
> Halo2_3.png: ?

Erase, no idea what they are for.

Lava: only texture "Lava_anim_0.png" is used on material file
Obsidian.png: erase
Treasury_specular.png: For Treasury.material -> Erase, specularity is handled now by material file

**Comment by akien-mga** [comment]
Thanks! I've dropped the mentioned textures then + the construct worker model, materials and textures.

**Comment by oyvindln** [comment]
If you want to cut the size further, you can try to run [OptiPNG](http://optipng.sourceforge.net/) or similar on the png files, which can potentially make them a bit smaller.

**Comment by Danimal696** [comment]
Pine-green seems unused

**Comment by akien-mga** [comment]
> Pine-green seems unused

According to my script it seems to be used in Bed.material:

```
pine_green.png: Wood    Bed.material
pine_green.png: WhiteCloth  Bed.material
pine_green.png: Blanket Bed.material
```

**Comment by akien-mga** [comment]
Fixed by #591.
